asynch/pool.py

```diff
@@ -294,35 +301,46 @@ async def _fill_with_connections(self, n: Optional[int] = None) -> None:
     async def connection(self) -> AsyncIterator[Connection]:
         """Get a connection from the pool.
 
-        :raises AsynchPoolError: if a connection cannot be acquired
-        :raises AsynchPoolError: if a connection cannot be released
+        If requested more connections than the pool can provide,
+        the pool gets blocked until a connection comes back.
+
+        :raises AsynchPoolError: if a connection cannot be acquired or released
 
         :return: a free connection from the pool
         :rtype: Connection
         """
 
-        async with self._lock:
-            if not self._free_connections:
-                to_create = min(self.minsize, self.maxsize - self.connections)
-                await self._fill_with_connections(to_create)
-            conn = await self._acquire_connection()
-        try:
-            yield conn
-        finally:
+        async with self._sem:
             async with self._lock:
-                await self._release_connection(conn)
+                if not self._free_connections:
+                    available_connections = self._maxsize - self.connections
+                    if available_connections < 0:
+                        msg = (
+                            f"the number of available connections ({available_connections}) "
```

Comment:
I guess it should be something like `f"the number of connections ({self.connections} ", because available_connections would be -1 or something, so doesn't exceed max_size & so on.